### PR TITLE
Hardhat launcher

### PR DESCRIPTION
**Pull request by fvictorio**
This is a proof of concept, but the code in `packages/hh` and the test setup can be reviewed (of course, there should be way more tests). The code in`cli/autocomplete.ts` is awful, ignore it for now.

How to test it:

- Build and globally install `packages/hh`
- `hh` won't do a lot: out of the box is pretty much an allias for `npx hardhat`
- Run `hh install-completion` and follow the instructions
- Go to some hardhat project and try the autocomplete

Things that are supported right now:

- Autocomplete of tasks and params, even if they were custom added. Params that are specific of a task should only be shown after some task is part of the command.
- Autocomplete of networks after `--network`

### CODE REVIEW

**Comment by alcuadrado** [comment]
🚀

> * Build and globally install `packages/hh`

How can I do this?

File: packages/hardhat-core/src/internal/cli/autocomplete.ts [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+  description?: string;
+}
+
+export async function complete({
+  line,
+  point,
+}: CompletionEnv): Promise<Suggestion[]> {
+  let hre: typeof hardhat;
+  try {
+    process.env.TS_NODE_TRANSPILE_ONLY = "1";
```
Comment (alcuadrado) [inline packages/hardhat-core/src/internal/cli/autocomplete.ts:22]: I'd move this to `hh`, and that way probably importing the lib would work.

**Review by alcuadrado — COMMENTED** [review]
This doesn't work correctly in Bash nor Fish, here are some screenshots:

Fish:

image
image
image

Bash:
image

**Comment by fvictorio** [comment]
> This doesn't work correctly in Bash nor Fish, here are some screenshots:

I just removed the descriptions from the suggestions. You can use them only in zsh by checking `process.env.SHELL`, but this is *very* unreliable, so I think it's best to just err on the side of caution for now.

File: packages/hardhat-core/src/internal/cli/ArgumentsParser.ts [hunk 2]
```diff
@@ -30,7 +30,10 @@ export class ArgumentsParser {
       });
     }
 
-    const parts = cLA.slice(ArgumentsParser.PARAM_PREFIX.length).split("-");
+    const parts = cLA
```
[the remark this answers is not in the document]
Comment (fvictorio) [reply packages/hardhat-core/src/internal/cli/ArgumentsParser.ts:33]: This method was failing for an incomplete flag like `--my-` (with a trailing `-`).

If you'd rather not modify this part of the code, we can just duplicate these methods in the `autocomplete` module, what do you think?

File: packages/hh/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      console.error(
+        "You are not inside a Hardhat project, or Hardhat is not locally installed"
+      );
+    } else {
+      console.error(`[hh] Unexpected error: ${e.message}`);
+    }
+    process.exit(1);
+  }
+
+  spawnSync("node", [pathToHardhat, ...process.argv.slice(2)], {
```
[the remark this answers is not in the document]
Comment (fvictorio) [reply packages/hh/src/index.ts]: Ugh, good catch.

File: packages/hh/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+  const { status } = spawnSync(
+    "node",
+    [pathToHardhat, ...process.argv.slice(2)],
+    {
+      stdio: "inherit",
+    }
+  );
+
+  process.exitCode = status ?? 0;
```
Comment (fvictorio) [inline packages/hh/src/index.ts:29]: @alcuadrado according to `spawnSync` docs, the status part of the result is:

> The exit code of the subprocess, or null if the subprocess terminated due to a signal.

I don't really know how to handle the `null` case. Maybe it should be 1 if it was interrupted?

File: packages/hardhat-core/src/internal/cli/autocomplete.ts (hunk 1 above)
Comment (fvictorio) [reply packages/hardhat-core/src/internal/cli/autocomplete.ts:17]: I tried this, but it stopped working with ts projects.

File: packages/hh/src/completion.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        }
+      );
+      const { complete } = require(pathToHardhatAutocomplete);
+      const suggestions = await complete(env);
+      return tabtab.log(suggestions);
+    } catch (e) {
+      return tabtab.log([]);
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (fvictorio) [reply packages/hh/src/completion.ts:44]: I added an error message.

I also removed the `uninstall` command. There seems to be a bug in `tabtab` that prevents it from working. We'll need to provide manual instructions on how to remove it.

**Pull request merged**